### PR TITLE
Israel playg 17 save recipes to database

### PR DESCRIPTION
**Pull request by Hazannovich**

### CODE REVIEW

File: daos/stored_recipes/test/stored_recipes_logic_test.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+def add_recipe(url, recipe):
+    if not stored_recipe_dao.url_in_db(url):
+        stored_recipe_dao.add_recipe_to_db(response=recipe)
+
+
+def get_recipe(url, ingred, instruct):
+    if stored_recipe_dao.url_in_db(url):
+        return stored_recipe_dao.get_recipe_from_db(url, ingred, instruct)
+
```
Comment (smartFreshBear) [inline daos/stored_recipes/test/stored_recipes_logic_test.py:129]: too much external to product logic, why do you need it? 
in unit test you need to test product components, what you have here is kinda mix between test logic and product logic

File: daos/stored_recipes/test/stored_recipes_logic_test.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+            "   3 בצלים",
+            "   1 פלפל ירוק חריף לא חובה אם לא רוצים חריפות עדינה",
+            "   5 קישואים",
+            "   1 קג דלעת/דלורית",
+            "   2 עגבניות",
+            "   חצי חבילה כוסברה",
+            "   56  עלי דפנה",
+            "   56 גרגרי פלפל אנגלי שלם",
+            "   מלח",
+            "   1 כפית פלפל שחור ",
```
Comment (smartFreshBear) [inline daos/stored_recipes/test/stored_recipes_logic_test.py:44]: nice idea, two things:
under the test module, add this mock recipe in a separate files-class, it's spamming the file.
the tested logic here is storing you don't need it to be so long (be cheap!)

File: app/app.py [hunk 3]
```diff
@@ -78,9 +87,16 @@ def find_recipe_in_url_window_algo_based():
...
     try:
+        if stored_recipes_dao.url_in_db(url):
+            return jsonify(stored_recipes_dao.get_recipe_from_db(url, ingredients, instructions))
         all_text, title = text_extractor.get_all_text_from_url(url=url)
         ingred_paragraph, instr_paragraph = window_key_word_based_algo.extract(ingredients, instructions, all_text)
-        return create_json_response(ingred_paragraph, instr_paragraph, title, url)
+        response = create_json_response(ingred_paragraph, instr_paragraph, title, url)
+
+        thread = threading.Thread(target=stored_recipes_dao.add_recipe_to_db, kwargs={"response": response})
+        thread.start()
```
Comment (smartFreshBear) [inline app/app.py:97]: I love your initiative here, but this is risky. 
it means that you will allocate 1 Thread per connection.
in  "rush hour' the server will collapse. 
either use correctly thread pool or just keep it simple (anyway it will happen only once per recipe)

**Review by smartFreshBear — CHANGES_REQUESTED** [review]
please refer to all comments

File: app/app.py [hunk 4]
```diff
@@ -78,9 +87,16 @@ def find_recipe_in_url_window_algo_based():
     instructions = request.form['instructions'].lower() == "true"
     ingredients = request.form['ingredients'].lower() == "true"
     try:
+        if stored_recipes_dao.url_in_db(url):
+            return jsonify(stored_recipes_dao.get_recipe_from_db(url, ingredients, instructions))
```
[the remark this answers is not in the document]
Comment (Hazannovich) [reply app/app.py:91]: it is working I am not sure I understand what you mean.

if url is in db then the user will get the recipe from the db with the get_recipe_from_db function.
else it will do an extraction as usual and at the end will store the extracted recipe in the db.

[the remark this answers is not in the document]
Comment (smartFreshBear) [reply app/app.py:91]: you right, got a bit confused cause github broke the line so I only see "return" disregard it.

File: daos/stored_recipes/test/stored_recipes_logic_test.py (hunk 1 above)
Comment (Hazannovich) [reply daos/stored_recipes/test/stored_recipes_logic_test.py:129]: yes it was a terrible choice, that builder design pattern does not belong in here.

File: app/app.py [hunk 5]
```diff
@@ -116,12 +133,12 @@ def en_find_recipe_in_url_window_algo_based():
...
         "ingredients": ["Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
-                            "Integer eu posuere massa. Mauris ut lectus feugiat,",
+                        "Integer eu posuere massa. Mauris ut lectus feugiat,",
                         "Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
                         "Integer eu posuere massa. Mauris ut lectus feugiat."],
         "instructions": ["Lorem ipsum dolor sit amet, consectetur adipiscing elit.",
-                            "Integer eu posuere massa. Mauris ut lectus feugiat,",
-                            " mattis est vel, finibus ligula. Maecenas euismod lacus non pellentesque tempus."],
+                         "Integer eu posuere massa. Mauris ut lectus feugiat,",
+                         " mattis est vel, finibus ligula. Maecenas euismod lacus non pellentesque tempus."],
```
Comment (smartFreshBear) [inline app/app.py:141]: is this relevant to the PR?

Comment (Hazannovich) [reply app/app.py]: no I changed it back it

File: daos/stored_recipes/test/stored_recipes_logic_test.py (hunk 2 above)
Comment (Hazannovich) [reply daos/stored_recipes/test/stored_recipes_logic_test.py]: Fixed!

File: app/app.py [hunk 6]
```diff
@@ -78,9 +87,17 @@ def find_recipe_in_url_window_algo_based():
...
+        if stored_recipes_dao.url_in_db(url):
+            return jsonify(stored_recipes_dao.get_recipe_from_db(url, ingredients, instructions))
         all_text, title = text_extractor.get_all_text_from_url(url=url)
         ingred_paragraph, instr_paragraph = window_key_word_based_algo.extract(ingredients, instructions, all_text)
-        return create_json_response(ingred_paragraph, instr_paragraph, title, url)
+        response = create_json_response(ingred_paragraph, instr_paragraph, title, url)
+        with ThreadPoolExecutor(max_workers=2) as executor:
+            future = executor.submit(stored_recipes_dao.add_recipe_to_db,
+                                     response=response)
+            future.result()
```
Comment (smartFreshBear) [inline app/app.py:98]: this about what happens here, we are creating a new ThreadPoolExcutor each time there is a call, how would you fix it?

Comment (smartFreshBear) [reply app/app.py:98]: I bet the flask itself has a solution for that.
the main idea is to manage ThreadPool outside of the scope of the "route"

File: daos/stored_recipes/test/stored_recipes_logic_test.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
+import unittest
+
+from flask import Flask
+import recipes_mocks
+from daos.database import DataBase
+from daos.stored_recipes import stored_recipes_manager
```
Comment (smartFreshBear) [inline daos/stored_recipes/test/stored_recipes_logic_test.py:6]: verify deletion in the test, and take advantage of it to make sure we are not creating garbage while testing. the database should be empty.

**Review by smartFreshBear — APPROVED** [review]
approved with two questions comments:
1- don't resolve conversation on GitHub, the one who start the thread will close it.
2- did you do a small sanity broken link didn't get any regression? (was't affected by this change?)

good work there!

**Comment by Hazannovich** [comment]
> 

1. LoL sorry 😅
2. Broken links related endpoints and table still work.
